maestro/trainer/models/florence_2/training.py

```diff
@@ -0,0 +1,348 @@
+            epoch_number=epoch + 1,
+            configuration=configuration,
+            optimizer=optimizer,
+            lr_scheduler=lr_scheduler,
+            checkpoints_leaderboard=checkpoints_leaderboard,
+            training_metrics_tracker=training_metrics_tracker,
+            validation_metrics_tracker=validation_metrics_tracker,
+        )
+
+
+def run_training_epoch(
+    processor: AutoProcessor,
+    model: PeftModel,
+    train_loader: DataLoader,
+    val_loader: Optional[DataLoader],
+    epoch_number: int,
+    configuration: TrainingConfiguration,
+    optimizer: Optimizer,
+    lr_scheduler: LRScheduler,
+    checkpoints_leaderboard: CheckpointsLeaderboard,
+    training_metrics_tracker: MetricsTracker,
+    validation_metrics_tracker: MetricsTracker,
+) -> None:
+    model.train()
+    training_losses: List[float] = []
+    training_iterator = tqdm(train_loader, desc=f"Epoch {epoch_number}/{configuration.training_epochs}")
+    for step_id, (inputs, answers) in enumerate(training_iterator):
+        input_ids = inputs["input_ids"]
+        pixel_values = inputs["pixel_values"]
+        labels = processor.tokenizer(
+            text=answers, return_tensors="pt", padding=True, return_token_type_ids=False
+        ).input_ids.to(configuration.device)
+        outputs = model(input_ids=input_ids, pixel_values=pixel_values, labels=labels)
+        loss = outputs.loss
+        loss.backward()
+        optimizer.step()
+        lr_scheduler.step()
+        optimizer.zero_grad()
+        loss = loss.item()
+        training_metrics_tracker.register(
+            metric="loss",
+            epoch=epoch_number,
+            step=step_id + 1,
+            value=loss,
+        )
+        training_losses.append(loss)
+        last_100_losses = training_losses[-100:]
+        loss_moving_average = sum(last_100_losses) / len(last_100_losses) if len(last_100_losses) > 0 else 0.0
+        training_iterator.set_description(
+            f"Epoch {epoch_number}/{configuration.training_epochs}. Loss: {round(loss_moving_average, 4)}"
+        )
+    if len(training_losses) > 0:
+        avg_train_loss = sum(training_losses) / len(training_losses)
+        print(f"Average Training Loss: {avg_train_loss}")
+    if val_loader is None or len(val_loader) == 0:
+        return None
+    validation_loss = run_validation_epoch(
+        processor=processor,
+        model=model,
+        loader=val_loader,
+        epoch_number=epoch_number,
+        configuration=configuration,
+    )
+    validation_metrics_tracker.register(
+        metric="loss",
+        epoch=epoch_number,
+        step=1,
+        value=validation_loss,
+    )
+    checkpoint_dir = os.path.join(configuration.training_dir, "checkpoints", str(epoch_number))
+    should_save, to_remove = checkpoints_leaderboard.register_checkpoint(
+        epoch=epoch_number,
+        path=checkpoint_dir,
+        loss=validation_loss,
+    )
+    if should_save:
```

Comment:
What do you think about a strategy similar to what YOLOv8 is doing, save the latest and best?

Reply:
👍